### PR TITLE
[FEATURE] Enables teachers to create their own adventures

### LINKED ISSUE #1515: [UI idea] Enable teachers to create assignments / adventures
**Idea incl level**
Currently with the class customizations implementation teachers are able to customize classes by selecting specific adventures for each class/level combination. As proposed by @OnnoEbbens in #954 it would be great if an assignment / adventure can be added in the class customization and shown on the tabs (maybe with a different border / color to differentiate?). On the longer term it would be nice if teachers can submit assignments to Hedy and we create an "assignment - overview" on which public assignments can be found and solved by all users.

### PR DESCRIPTION
**Pull request by TiBiBa**
**Description**
In this PR we implement a new feature enabling teachers to create, edit and delete their own adventures! We add a new section to the `for-teachers` page called "My adventures". Teachers can create a new adventure and are able to select a corresponding Hedy level and use commands and code snippets as well. Created adventures can be implemented in the adventures tab of classes under "class customizations".

**Fix for**
This PR fixes #1515.

**How to test**
Due to the complexity and large amount of related files this need quite some testing! Make sure you are a teacher and navigate to the teachers account and notice the new section. Create an adventure, alter it, delete it or select any other option. Then make sure you have a class and navigate to class customizations. Notice that the created level is now visible as an option in the corresponding Hedy level. Select it, login as a student (or join your own class) and navigate to that specific level. Notice that the adventure can be found within the adventure tab in a distinct color.

**Screenshots**
Schermafbeelding 2022-01-27 om 15 00 57
Schermafbeelding 2022-01-27 om 15 00 49
Schermafbeelding 2022-01-27 om 16 55 16
Schermafbeelding 2022-01-27 om 16 55 35
Schermafbeelding 2022-01-27 om 16 55 49

### CODE REVIEW

**Comment by TiBiBa** [comment]
Schermafbeelding 2022-01-26 om 13 07 15

**Comment by TiBiBa** [comment]
Notes to self:

- Fix re-direct delete button on customize page
- Delete (or make useful) -> viewing page (redundant now preview modal works)
- Convert texts to .yamls instead of hard-coding
- Make URL usage on back-end consistent
- Implement achievements within class customizations
- Fix iteration on `incl-adventure-tabs.html`
- Give own adventures different color

**Comment by TiBiBa** [comment]
Hi! This PR is almost ready,  but due to the complexity it would be really nice to get some input on missing features, bugs or unclear UI. Can you take a look? @MarleenGilsing @OnnoEbbens

**Comment by MarleenGilsing** [comment]
It looks amazing, Timon! Great job!
It was really easy to add an adventure. I loved the example function, so you can check what it looks like. 
And it was easy to add the adventure to your class as well. 

Makes me think of another function that might be usefull for teachers (sorry for the werkverschaffing haha). It would be nice for teachers to have a button 'student mode' so they know what the interface looks like for their students. Obviously they could also just make a student-account and enter it to their class to see that, but this might be a nice touch.

I have only one issue with your interface atm and that is the saving of the program. After saving it you stay on the same page, which seems weird. It feels like you should be sent back to the teacherpage, because now I keep saving it 3 times 'just in case'. Also the 'your adventure is saved' message is red, which gives off the vibe that something went wrong. Maybe make it green?

Some other tiny things I noticed:
- The message 'vul een avonturen naam in' 'loopt niet echt lekker'. Maybe change it to 'Geef je nieuwe avontuur een naam' or something. 
- Lerarenpagina is one word, in the back-button it's two. 
- When you make a new adventure and save it, it asks if you want to update it, which might be a bit weird because I have never saved it before. Not that I think this would cause any real problems...
- When pressing verwijder you get an English message. 
- The class page doesn't work for me atm. But that might very well be a problem with my computer. 
- When I tried to save the program in the English I got a Dutch notification 'Vink dit selectievakje aan als je wil doorgaan' (again might be me)

But again, overall I loved it!

**Comment by TiBiBa** [comment]
> Makes me think of another function that might be usefull for teachers (sorry for the werkverschaffing haha). It would be nice for teachers to have a button 'student mode' so they know what the interface looks like for their students. Obviously they could also just make a student-account and enter it to their class to see that, but this might be a nice touch.

Will take a look at this, should be too hard!

> I have only one issue with your interface atm and that is the saving of the program. After saving it you stay on the same page, which seems weird. It feels like you should be sent back to the teacherpage, because now I keep saving it 3 times 'just in case'. Also the 'your adventure is saved' message is red, which gives off the vibe that something went wrong. Maybe make it green?

Good point! Will take a look at this. The issue there is that we would want to provide the user with feedback if the updating was successful. When re-directing to the teacher page this is harder to manage. But, if you think the re-direct is feedback enough I will implement this. (There is also a last-update timestamp on the teacher page to provide some info) The red message was a known bug, I think I fixed it with a commit somewhere yesterday; but will take a look at this as well just to be sure.

> Some other tiny things I noticed:
> 
> * The message 'vul een avonturen naam in' 'loopt niet echt lekker'. Maybe change it to 'Geef je nieuwe avontuur een naam' or something.

Good catch! Will change this.

> * Lerarenpagina is one word, in the back-button it's two.

Will change this as well!

> * When you make a new adventure and save it, it asks if you want to update it, which might be a bit weird because I have never saved it before. Not that I think this would cause any real problems...

Hmm, yes we currently use one function for this. But will add an extra check to verify if the adventure is "new". In that case don't give this extra verification.

> * When pressing verwijder you get an English message.

My bad! (Forgot a yaml somewhere) Will fix!

> * The class page doesn't work for me atm. But that might very well be a problem with my computer.

Weird! This should work, but will take a look as well.

> * When I tried to save the program in the English I got a Dutch notification 'Vink dit selectievakje aan als je wil doorgaan' (again might be me)

This is due to your browser language. I set the checkbox as "required", then the browser itself makes sure that you can't submit any data without checking it. But, we don't have influence on the feedback message of the browser. So dependent on your browser and browser language a message is shown.

**Comment by TiBiBa** [comment]
Small fixes:

- ~~Fix dutch yaml on name pop-up~~
- ~~Fix 'lerarenpagina'~~
- ~~Don't show update pop-up on first edit~~
- Implement dutch message on delete
- ~~Fix red update message~~

**Comment by TiBiBa** [comment]
All fixed! I think we are ready for merge!

**Comment by TiBiBa** [comment]
**Before merging: create corresponding tables on alpha and beta**

**Review by boryanagoncharenko — APPROVED** [review]

**Review by ThomasKluiters — COMMENTED** [review]
Awesome job!

Some minor comments / questions - will do some testing as well!

**Comment by ThomasKluiters** [comment]
A general remark I have on this feature is that it may be a good idea to purify the adventure 'code' using something like https://github.com/cure53/DOMPurify. Currently, I'm free to create an adventure with <script> tags and perform arbitrary JavaScript code by sharing adventures.

I do not think that this poses a very serious threat to Hedy users, and from a pragmatic standpoint this behaviour is not too concerning, though, in my personal opinion, I think we should leverage some library to "clean" the submitted 'code' to protect users.

**Comment by Felienne** [comment]
> A general remark I have on this feature is that it may be a good idea to purify the adventure 'code' using something like https://github.com/cure53/DOMPurify. Currently, I'm free to create an adventure with <script> tags and perform arbitrary JavaScript code by sharing adventures.
> 
> I do not think that this poses a very serious threat to Hedy users, and from a pragmatic standpoint this behaviour is not too concerning, though, in my personal opinion, I think we should leverage some library to "clean" the submitted 'code' to protect users.

Yeah I do agree this is a good idea, it is sad but people on the internet aren't always well meaning sadly :(

File: static/js/teachers.ts
Comment (TiBiBa) [reply static/js/teachers.ts:337]: It is! A reload is not necessary here as the front-end is already up-to-date with the changes. I thought to leave it this way until I have time to improve the whole class customization. But will add a comment to clarify.

File: static/js/modal.ts
Comment (TiBiBa) [reply static/js/modal.ts:84]: Good catch! Removed them.

File: static/js/app.ts
Comment (TiBiBa) [reply static/js/app.ts:50]: Nope, but was a mistake I noticed when implemented this (and using the same functionality). Thought I wasn't worth creating a dedicated issue.

File: static/js/teachers.ts
Comment (TiBiBa) [reply static/js/teachers.ts:189]: Not sure what you mean. To prevent teachers from accidentally creation adventures with a whitespace as name?

File: website/database.py
Comment (TiBiBa) [reply website/database.py:371]: Fixed

File: website/teacher.py
Comment (TiBiBa) [reply website/teacher.py:403]: Fixed

**Comment by TiBiBa** [comment]
> A general remark I have on this feature is that it may be a good idea to purify the adventure 'code' using something like https://github.com/cure53/DOMPurify. Currently, I'm free to create an adventure with <script> tags and perform arbitrary JavaScript code by sharing adventures.
> 
> I do not think that this poses a very serious threat to Hedy users, and from a pragmatic standpoint this behaviour is not too concerning, though, in my personal opinion, I think we should leverage some library to "clean" the submitted 'code' to protect users.

Not sure this will happen often as teacher accounts are set manually and we have a clear control on the people able to create adventures. But it doesn't have any negative effect. Will look into this.

**Comment by TiBiBa** [comment]
> A general remark I have on this feature is that it may be a good idea to purify the adventure 'code' using something like https://github.com/cure53/DOMPurify. Currently, I'm free to create an adventure with <script> tags and perform arbitrary JavaScript code by sharing adventures.
> 
> I do not think that this poses a very serious threat to Hedy users, and from a pragmatic standpoint this behaviour is not too concerning, though, in my personal opinion, I think we should leverage some library to "clean" the submitted 'code' to protect users.

We also might have to expand on this topic. As we store the adventure as plain text and parse it to `|commonmark` with jinja **all** HTML attributes are shown. Users are able to include headers, iframe, re-direct links and more. So if we want to purify the code we have to keep all these situations in mind. However, I think this is a complex task and as only teachers are able to create adventures we should be fine. I think we can implement this in three ways:

1. Do nothing, assume this is going to be fine
2. The admin verifies adventures before teachers can use them
3.

**Comment by TiBiBa** [comment]
Hi @ThomasKluiters I just implemented the `DOMPurify` library. Would be create if you could check it out! I did some testing and it all seems to working fine with sanitizing our textarea input.

**Review by ThomasKluiters — APPROVED** [review]
Awesome job Timon!

**Comment by ThomasKluiters** [comment]
Sorry for the later responses - was quite occupied :(

**Comment by Felienne** [comment]
> Sorry for the later responses - was quite occupied :(

No problem, we are superhappy with your help!!

File: static/js/teachers.ts
Comment (Felienne) [reply static/js/teachers.ts:189]: yeah I think he means with whitespace at the end, like "adventure " Do we want to so something against that?

Comment (TiBiBa) [reply static/js/teachers.ts:189]: Will fix!

Comment (TiBiBa) [reply static/js/teachers.ts:189]: We do have a similar structure on multiple locations, such as re-naming a class. Will take a look at these situations to unify the input handling later on.

**Comment by Felienne** [comment]
Hi @TiBiBa!

I have one little request but we can do it separately as well to merge this. I think we want to allow teachers to make adventures just for their own class, so the tick mark should not be needed to proceed, it should be optional.

image

**Comment by Felienne** [comment]
Also, if I click opslaan, I do not get a modal or popup showing me that the adventure is saved, is that as intended?

**Comment by TiBiBa** [comment]
> Hi @TiBiBa!
> 
> I have one little request but we can do it separately as well to merge this. I think we want to allow teachers to make adventures just for their own class, so the tick mark should not be needed to proceed, it should be optional.
> 
> image

I can fix this on this PR! We do have to store some additional value in the db than such as 'public' to track their choice.

**Comment by TiBiBa** [comment]
> Also, if I click opslaan, I do not get a modal or popup showing me that the adventure is saved, is that as intended?

Weird! I should return a modal.alert(), will take a look.

**Comment by TiBiBa** [comment]
> Also, if I click opslaan, I do not get a modal or popup showing me that the adventure is saved, is that as intended?

It does seem to work fine for me!
Schermafbeelding 2022-02-04 om 11 24 42

**Comment by Felienne** [comment]
Nice, this now all works well!! And I love the code blocks that can be copied as the regular ones!

**Review by Felienne — APPROVED** [review]

**Pull request merged**
